File: libs/document-processor-next/src/engine/contract-bundle.ts
import { blueIds } from '@blue-repository/core';

import { KEY_CHECKPOINT } from '../constants/processor-contract-constants.js';
import type {
  ChannelContract,
  HandlerContract,
  MarkerContract,
  ProcessEmbeddedMarker,
  ChannelEventCheckpoint,
} from '../model/index.js';

type StoredChannel = {
  readonly key: string;
  readonly contract: ChannelContract;
  readonly blueId: string;
};

type StoredMarker = {
  readonly key: string;
  readonly contract: MarkerContract;
  readonly blueId: string;
};

type StoredHandler = {
  readonly key: string;
  readonly contract: HandlerContract;
  readonly blueId: string;
};

const CHANNEL_EVENT_CHECKPOINT_BLUE_ID = blueIds['Channel Event Checkpoint'];

function contractOrder(contract: { readonly order?: number | null }): number {
  return typeof contract.order === 'number' ? contract.order : 0;
}

export class ChannelBinding {
  constructor(
    private readonly bindingKey: string,
    private readonly bindingContract: ChannelContract,
    private readonly bindingBlueId: string
  ) {}

  key(): string {
    return this.bindingKey;
  }

  contract(): ChannelContract {
    return this.bindingContract;
  }

  blueId(): string {
    return this.bindingBlueId;
  }

  order(): number {
    return contractOrder(this.bindingContract);
  }
}

export class HandlerBinding {
  constructor(
    private readonly bindingKey: string,
    private readonly bindingContract: HandlerContract,
    private readonly bindingBlueId: string
  ) {}

  key(): string {
    return this.bindingKey;
  }

  contract(): HandlerContract {
    return this.bindingContract;
  }

  blueId(): string {
    return this.bindingBlueId;
  }

  order(): number {
    return contractOrder(this.bindingContract);
  }
}

export class ContractBundle {
  private checkpointDeclared: boolean;

  private constructor(
    private readonly channels: Map<string, StoredChannel>,
    private readonly handlersByChannel: Map<string, StoredHandler[]>,
    private readonly markerStore: Map<string, StoredMarker>,
    private readonly embeddedPathsInternal: readonly string[],
    checkpointDeclared: boolean
  ) {
    this.checkpointDeclared = checkpointDeclared;
  }

  static fromComponents(
    channels: Map<string, StoredChannel>,
    handlersByChannel: Map<string, StoredHandler[]>,
    markerStore: Map<string, StoredMarker>,
    embeddedPaths: readonly string[],
    checkpointDeclared: boolean
  ): ContractBundle {
    return new ContractBundle(
      channels,
      handlersByChannel,
      markerStore,
      embeddedPaths,
      checkpointDeclared
    );
  }

  static builder(): ContractBundleBuilder {
    return new ContractBundleBuilder();
  }

  static empty(): ContractBundle {
    return ContractBundle.builder().build();
  }

  markers(): Map<string, MarkerContract> {
    return new Map(
      Array.from(this.markerStore.entries(), ([key, entry]) => [
        key,
        entry.contract,
      ])
    );
  }

  marker(key: string): MarkerContract | undefined {
    return this.markerStore.get(key)?.contract;
  }

  markerEntries(): Array<[string, MarkerContract]> {
    return Array.from(this.markerStore.entries(), ([key, entry]) => [
      key,
      entry.contract,
    ]);
  }

  embeddedPaths(): readonly string[] {
    return this.embeddedPathsInternal;
  }

  hasCheckpoint(): boolean {
    return this.checkpointDeclared;
  }

  registerCheckpointMarker(checkpoint: ChannelEventCheckpoint): void {
    if (this.checkpointDeclared) {
      throw new Error(
        'Duplicate Channel Event Checkpoint markers detected in same contracts map'
      );
    }
    this.markerStore.set(KEY_CHECKPOINT, {
      key: KEY_CHECKPOINT,
      contract: checkpoint,
      blueId: CHANNEL_EVENT_CHECKPOINT_BLUE_ID,
    });
    this.checkpointDeclared = true;
  }

  handlersFor(channelKey: string): HandlerBinding[] {
    const handlers = this.handlersByChannel.get(channelKey);
    if (!handlers || handlers.length === 0) {
      return [];
    }
    return [...handlers]
      .map(
        (entry) => new HandlerBinding(entry.key, entry.contract, entry.blueId)
      )
      .sort((a, b) => {
        const orderDiff = a.order() - b.order();
        if (orderDiff !== 0) {
          return orderDiff;
        }
        return a.key().localeCompare(b.key());
      });
  }

  channelsOfType(...blueIds: readonly string[]): ChannelBinding[] {
    const filter = blueIds.length > 0 ? new Set(blueIds) : null;
    const bindings = Array.from(this.channels.values())
      .filter((entry) => !filter || filter.has(entry.blueId))
      .map(
        (entry) => new ChannelBinding(entry.key, entry.contract, entry.blueId)
      );
    bindings.sort((a, b) => {
      const orderDiff = a.order() - b.order();
      if (orderDiff !== 0) {
        return orderDiff;
      }
      return a.key().localeCompare(b.key());
    });
    return bindings;
  }
}

export class ContractBundleBuilder {
  private readonly channels = new Map<string, StoredChannel>();
  private readonly handlersByChannel = new Map<string, StoredHandler[]>();
  private readonly markerStore = new Map<string, StoredMarker>();
  private embeddedPaths: string[] = [];
  private embeddedDeclared = false;
  private checkpointDeclared = false;

  addChannel(key: string, contract: ChannelContract, blueId: string): this {
    this.channels.set(key, { key, contract, blueId });
    return this;
  }

  addHandler(key: string, contract: HandlerContract, blueId: string): this {
    const channelKey = contract.channel;
    if (!channelKey) {
      throw new Error(`Handler ${key} must declare channel`);
    }
    const list = this.handlersByChannel.get(channelKey) ?? [];
    list.push({ key, contract, blueId });
    this.handlersByChannel.set(channelKey, list);
    return this;
  }

  setEmbedded(embedded: ProcessEmbeddedMarker): this {
    if (this.embeddedDeclared) {
      throw new Error(
        'Multiple Process Embedded markers detected in same contracts map'
      );
    }
    this.embeddedDeclared = true;
    this.embeddedPaths = embedded.paths ? [...embedded.paths] : [];
    return this;
  }

  addMarker(key: string, contract: MarkerContract, blueId: string): this {
    if (key === KEY_CHECKPOINT && blueId !== CHANNEL_EVENT_CHECKPOINT_BLUE_ID) {
      throw new Error(
        "Reserved key 'checkpoint' must contain a Channel Event Checkpoint"
      );
    }
    if (blueId === CHANNEL_EVENT_CHECKPOINT_BLUE_ID) {
      if (key !== KEY_CHECKPOINT) {
        throw new Error(
          `Channel Event Checkpoint must use reserved key 'checkpoint' at key '${key}'`
        );
      }
      if (this.checkpointDeclared) {
        throw new Error(
          'Duplicate Channel Event Checkpoint markers detected in same contracts map'
        );
      }
      this.checkpointDeclared = true;
    }
    this.markerStore.set(key, { key, contract, blueId });
    return this;
  }

  build(): ContractBundle {
    return ContractBundle.fromComponents(
      new Map(this.channels),
      new Map(
        Array.from(this.handlersByChannel.entries(), ([key, list]) => [
          key,
          [...list],
        ])
      ),
      new Map(this.markerStore),
      [...this.embeddedPaths],
      this.checkpointDeclared
    );
  }
}

File: references/java-sdk/src/main/java/blue/language/processor/ContractBundle.java
package blue.language.processor;

import blue.language.processor.model.ChannelContract;
import blue.language.processor.model.HandlerContract;
import blue.language.processor.model.MarkerContract;
import blue.language.processor.model.ProcessEmbedded;
import blue.language.processor.model.ChannelEventCheckpoint;
import blue.language.processor.util.ProcessorContractConstants;
import blue.language.processor.util.PointerUtils;
import blue.language.model.TypeBlueId;

import java.util.ArrayList;
import java.util.Collections;
import java.util.Comparator;
import java.util.LinkedHashMap;
import java.util.LinkedHashSet;
import java.util.List;
import java.util.Map;
import java.util.Set;

/**
 * Collection of contracts bound to a scope, along with helper accessors.
 */
public final class ContractBundle {

    private final Map<String, ChannelContract> channels;
    private final Map<String, List<HandlerBinding>> handlersByChannel;
    private final Map<String, MarkerContract> markers;
    private final List<String> embeddedPaths;
    private boolean checkpointDeclared;

    private final Map<String, MarkerContract> markersView;
    private final List<String> embeddedPathsView;

    private ContractBundle(Map<String, ChannelContract> channels,
                           Map<String, List<HandlerBinding>> handlersByChannel,
                           Map<String, MarkerContract> markers,
                           List<String> embeddedPaths,
                           boolean checkpointDeclared) {
        this.channels = channels;
        this.handlersByChannel = handlersByChannel;
        this.markers = markers;
        this.embeddedPaths = embeddedPaths;
        this.checkpointDeclared = checkpointDeclared;

        this.markersView = Collections.unmodifiableMap(this.markers);
        this.embeddedPathsView = Collections.unmodifiableList(this.embeddedPaths);
    }

    public static Builder builder() {
        return new Builder();
    }

    public static ContractBundle empty() {
        return builder().build();
    }

    public Map<String, MarkerContract> markers() {
        return markersView;
    }

    public MarkerContract marker(String key) {
        return markers.get(key);
    }

    public Set<Map.Entry<String, MarkerContract>> markerEntries() {
        return Collections.unmodifiableSet(new LinkedHashSet<>(markers.entrySet()));
    }

    public List<String> embeddedPaths() {
        return embeddedPathsView;
    }

    public boolean hasCheckpoint() {
        return checkpointDeclared;
    }

    public void registerCheckpointMarker(ChannelEventCheckpoint checkpoint) {
        if (checkpointDeclared) {
            throw new IllegalStateException("Duplicate Channel Event Checkpoint markers detected in same contracts map");
        }
        markers.put(ProcessorContractConstants.KEY_CHECKPOINT, checkpoint);
        checkpointDeclared = true;
    }

    public List<HandlerBinding> handlersFor(String channelKey) {
        List<HandlerBinding> handlers = handlersByChannel.get(channelKey);
        if (handlers == null || handlers.isEmpty()) {
            return Collections.emptyList();
        }
        List<HandlerBinding> sorted = new ArrayList<>(handlers);
        sorted.sort(Comparator
                .comparingInt(HandlerBinding::order)
                .thenComparing(HandlerBinding::key));
        return sorted;
    }

    public ChannelBinding channel(String key) {
        if (key == null) {
            return null;
        }
        String normalized = key.trim();
        if (normalized.isEmpty()) {
            return null;
        }
        ChannelContract contract = channels.get(normalized);
        if (contract == null) {
            return null;
        }
        return new ChannelBinding(normalized, contract);
    }

    public List<ChannelBinding> channelsOfType(Class<? extends ChannelContract> type) {
        List<ChannelBinding> result = new ArrayList<>();
        for (Map.Entry<String, ChannelContract> entry : channels.entrySet()) {
            ChannelContract contract = entry.getValue();
            if (type.isInstance(contract)) {
                result.add(new ChannelBinding(entry.getKey(), contract));
            }
        }
        result.sort(Comparator
                .comparingInt(ChannelBinding::order)
                .thenComparing(ChannelBinding::key));
        return result;
    }

    public List<ChannelBinding> channelsOfType(String blueId) {
        if (blueId == null || blueId.trim().isEmpty()) {
            return Collections.emptyList();
        }
        String normalized = blueId.trim();
        List<ChannelBinding> result = new ArrayList<>();
        for (Map.Entry<String, ChannelContract> entry : channels.entrySet()) {
            ChannelContract contract = entry.getValue();
            if (channelMatchesBlueId(contract, normalized)) {
                result.add(new ChannelBinding(entry.getKey(), contract));
            }
        }
        result.sort(Comparator
                .comparingInt(ChannelBinding::order)
                .thenComparing(ChannelBinding::key));
        return result;
    }

    public static final class ChannelBinding {
        private final String key;
        private final ChannelContract contract;

        ChannelBinding(String key, ChannelContract contract) {
            this.key = key;
            this.contract = contract;
        }

        public String key() {
            return key;
        }

        public ChannelContract contract() {
            return contract;
        }

        public int order() {
            Integer order = contract.getOrder();
            return order != null ? order : 0;
        }
    }

    public static final class HandlerBinding {
        private final String key;
        private final HandlerContract contract;

        HandlerBinding(String key, HandlerContract contract) {
            this.key = key;
            this.contract = contract;
        }

        public String key() {
            return key;
        }

        public HandlerContract contract() {
            return contract;
        }

        public int order() {
            Integer order = contract.getOrder();
            return order != null ? order : 0;
        }
    }

    public static final class Builder {
        private final Map<String, ChannelContract> channels = new LinkedHashMap<>();
        private final Map<String, List<HandlerBinding>> handlersByChannel = new LinkedHashMap<>();
        private final Map<String, MarkerContract> markers = new LinkedHashMap<>();
        private final Set<String> usedContractKeys = new LinkedHashSet<>();
        private final List<String> embeddedPaths = new ArrayList<>();
        private boolean embeddedDeclared;
        private boolean checkpointDeclared;

        private Builder() {
        }

        public Builder addChannel(String key, ChannelContract contract) {
            String normalizedKey = validateContractKey(key, "Channel");
            ensureUniqueContractKey(normalizedKey);
            channels.put(normalizedKey, contract);
            return this;
        }

        public Builder addHandler(String key, HandlerContract contract) {
            String normalizedKey = validateContractKey(key, "Handler");
            ensureUniqueContractKey(normalizedKey);
            String channelKey = normalizeChannelKey(contract.getChannelKey(), normalizedKey);
            contract.setChannelKey(channelKey);
            handlersByChannel
                    .computeIfAbsent(channelKey, k -> new ArrayList<>())
                    .add(new HandlerBinding(normalizedKey, contract));
            return this;
        }

        public Builder setEmbedded(ProcessEmbedded embedded) {
            if (embeddedDeclared) {
                throw new IllegalStateException("Multiple Process Embedded markers detected in same contracts map");
            }
            embeddedDeclared = true;
            if (embedded.getPaths() != null) {
                embeddedPaths.clear();
                Set<String> normalizedPaths = new LinkedHashSet<>();
                for (String path : embedded.getPaths()) {
                    if (path == null) {
                        continue;
                    }
                    String trimmed = path.trim();
                    if (trimmed.isEmpty()) {
                        continue;
                    }
                    normalizedPaths.add(PointerUtils.normalizeRequiredPointer(trimmed, "Embedded path"));
                }
                embeddedPaths.addAll(normalizedPaths);
            }
            return this;
        }

        public Builder addMarker(String key, MarkerContract contract) {
            String normalizedKey = validateContractKey(key, "Marker");
            if (contract instanceof ChannelEventCheckpoint) {
                if (!ProcessorContractConstants.KEY_CHECKPOINT.equals(normalizedKey)) {
                    throw new IllegalStateException(
                            "Channel Event Checkpoint must use reserved key 'checkpoint' at key '" + normalizedKey + "'");
                }
                if (checkpointDeclared) {
                    throw new IllegalStateException("Duplicate Channel Event Checkpoint markers detected in same contracts map");
                }
            } else if (ProcessorContractConstants.KEY_CHECKPOINT.equals(normalizedKey)) {
                throw new IllegalStateException(
                        "Reserved key 'checkpoint' must contain a Channel Event Checkpoint");
            }
            ensureUniqueContractKey(normalizedKey);
            if (contract instanceof ChannelEventCheckpoint) {
                checkpointDeclared = true;
            }
            markers.put(normalizedKey, contract);
            return this;
        }

        public ContractBundle build() {
            for (String channelKey : handlersByChannel.keySet()) {
                if (!channels.containsKey(channelKey)) {
                    throw new IllegalStateException("Handler references missing channel: " + channelKey);
                }
            }
            return new ContractBundle(channels, handlersByChannel, markers, embeddedPaths, checkpointDeclared);
        }

        private String validateContractKey(String key, String contractKind) {
            if (key == null || key.trim().isEmpty()) {
                throw new IllegalStateException(contractKind + " contract key must not be blank");
            }
            return key.trim();
        }

        private void ensureUniqueContractKey(String key) {
            if (!usedContractKeys.add(key)) {
                throw new IllegalStateException("Duplicate contract key: " + key);
            }
        }

        private String normalizeChannelKey(String channelKey, String handlerKey) {
            if (channelKey == null) {
                throw new IllegalStateException("Handler " + handlerKey + " must declare channel");
            }
            String normalized = channelKey.trim();
            if (normalized.isEmpty()) {
                throw new IllegalStateException("Handler " + handlerKey + " must declare channel");
            }
            return normalized;
        }
    }

    private static boolean channelMatchesBlueId(ChannelContract contract, String blueId) {
        if (contract == null || blueId == null) {
            return false;
        }
        TypeBlueId annotation = contract.getClass().getAnnotation(TypeBlueId.class);
        if (annotation == null) {
            return false;
        }
        String[] values = annotation.value();
        if (values != null) {
            for (String value : values) {
                if (blueId.equals(value != null ? value.trim() : null)) {
                    return true;
                }
            }
        }
        String defaultValue = annotation.defaultValue();
        return blueId.equals(defaultValue != null ? defaultValue.trim() : null);
    }
}
